File: common/src/main/kotlin/net/impleri/playerskills/events/SkillChangedEvent.kt
package net.impleri.playerskills.events

import dev.architectury.event.Event
import dev.architectury.event.EventFactory
import net.impleri.playerskills.api.Skill
import net.impleri.playerskills.api.SkillType
import net.minecraft.world.entity.player.Player
import java.util.function.Consumer

class SkillChangedEvent<T>(val player: Player, val skill: Skill<T>, val previous: Skill<T>) {
  val type: SkillType<T>? = SkillType.find(skill)

  companion object {
    val EVENT: Event<Consumer<SkillChangedEvent<*>>> = EventFactory.createConsumerLoop()
  }
}

File: common/src/main/scala/net/impleri/playerskills/events/SkillChangedEvent.scala
package net.impleri.playerskills.events

import net.impleri.playerskills.api.skills.Skill
import net.impleri.slab.entity.Player

case class SkillChangedEvent[T](
  player: Player,
  next: Option[Skill[T]],
  previous: Option[Skill[T]],
)
